File: training-data/php/hash_set_open_addressing.php
<?php

/** Open-addressing string set with linear probing and tombstones. */
final class OpenHashSet implements Countable
{
    private const TOMBSTONE = "\0deleted\0";
    private array $slots;
    private int $size = 0;
    private int $used = 0;

    public function __construct(int $capacity = 8)
    {
        $this->slots = array_fill(0, $capacity, null);
    }

    private function index(string $key): int
    {
        return crc32($key) % count($this->slots);
    }

    private function find(string $key): ?int
    {
        $cap = count($this->slots);
        for ($i = $this->index($key), $n = 0; $n < $cap; $i = ($i + 1) % $cap, $n++) {
            if ($this->slots[$i] === null) {
                return null;
            }
            if ($this->slots[$i] === $key) {
                return $i;
            }
        }
        return null;
    }

    public function contains(string $key): bool
    {
        return $this->find($key) !== null;
    }

    public function add(string $key): bool
    {
        if ($this->contains($key)) {
            return false;
        }
        if (($this->used + 1) * 2 > count($this->slots)) {
            $live = array_filter($this->slots, fn ($s) => $s !== null && $s !== self::TOMBSTONE);
            $this->slots = array_fill(0, count($this->slots) * 2, null);
            $this->size = $this->used = 0;
            foreach ($live as $k) {
                $this->add($k);
            }
        }
        $i = $this->index($key);
        while ($this->slots[$i] !== null && $this->slots[$i] !== self::TOMBSTONE) {
            $i = ($i + 1) % count($this->slots);
        }
        if ($this->slots[$i] === null) {
            $this->used++;
        }
        $this->slots[$i] = $key;
        $this->size++;
        return true;
    }

    public function remove(string $key): bool
    {
        $i = $this->find($key);
        if ($i === null) {
            return false;
        }
        $this->slots[$i] = self::TOMBSTONE;
        $this->size--;
        return true;
    }

    public function count(): int
    {
        return $this->size;
    }
}

$set = new OpenHashSet();
foreach (explode(' ', 'the quick brown fox jumps over the lazy dog') as $w) {
    $set->add($w);
}
$set->remove('fox');
var_export([count($set), $set->contains('fox'), $set->contains('dog')]);
echo PHP_EOL;

File: training-data/ruby/hash_set_open_addressing.rb
class OpenHashSet
  TOMBSTONE = Object.new.freeze

  attr_reader :size

  def initialize
    @slots = Array.new(8)
    @size = 0
    @used = 0
  end

  def add(key)
    return self if include?(key)

    rehash if (@used + 1) * 2 > @slots.size
    i = key.hash % @slots.size
    i = (i + 1) % @slots.size while @slots[i] && !@slots[i].equal?(TOMBSTONE)
    @used += 1 if @slots[i].nil?
    @slots[i] = key
    @size += 1
    self
  end
  alias << add

  def include?(key) = !find(key).nil?

  def delete(key)
    i = find(key)
    return false unless i

    @slots[i] = TOMBSTONE
    @size -= 1
    true
  end

  private

  def find(key)
    i = key.hash % @slots.size
    @slots.size.times do
      slot = @slots[i]
      return nil if slot.nil?
      return i if !slot.equal?(TOMBSTONE) && slot.eql?(key)

      i = (i + 1) % @slots.size
    end
    nil
  end

  def rehash
    live = @slots.reject { |s| s.nil? || s.equal?(TOMBSTONE) }
    @slots = Array.new(@slots.size * 2)
    @size = @used = 0
    live.each { |k| add(k) }
  end
end

set = OpenHashSet.new
'the quick brown fox jumps over the lazy dog'.split.each { |w| set << w }
set.delete('fox')
p [set.size, set.include?('fox'), set.include?('dog')]
